### ISSUE TITLE
Redesign dataloader re-instantiation context managers

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/lightning#15179]
## Proposed refactor

Redesign the context managers that enable re-instantiations of DataLoader after #14992 deprecates the run method design.

### Motivation

We currently have two context managers over the run method:

https://github.com/Lightning-AI/lightning/blob/776432fd7e34cc02e9dbfb16ad8ec8491c176dbf/src/lightning_lite/lite.py#L394-L397

- _replace_dunder_methods(DataLoader, "dataset")
- _replace_dunder_methods(BatchSampler)

These get applied for the user if they implement the run method. However, this won't be the case anymore after #14992. We need to find an alternative approach.

### Pitch

We probably need to call these methods at the start of the program, in `LightningLite.__init__`. But it is unclear how to undo the patching which is usually done in `__exit__`. 

cc @carmocca @otaj @justusschock  

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

**Comment by carmocca** [comment]
Can't we simply move them to `LightningLite.setup_dataloaders`?

**Comment by awaelchli** [comment]
I thought that we need to patch the class _before_ the user instantiates a dataloader? Otherwise why would we need such a complicated re-instantiation logic?

**Comment by carmocca** [comment]
Right. Then we might not want to undo the patching at all. It's the same design issue of "when should we teardown DDP" and such.

**Comment by otaj** [comment]
I think we need to adapt it completely. I don't think we necessarily care about unpatching, we just need to be sure that on re-instantiation we are calling original `__init__` and not the patched one (which is taken care of automatically, since we are unpatching the method)

**Comment by carmocca** [comment]
The caveat is a feature that we definitely want to support. This means that we do it at import time or ask the user to do it through lite `dataloder = LightningLite.dataloader(...)`

The former is the "magic" way of course. But for that we should make sure it's failsafe, idempotent, and that giving up the "un-patch"ing part is okay.

**Comment by awaelchli** [comment]
What's annoying me right now is that, since PL depends on Lite, the patching gets applied globally on import for PL too, and I can't avoid that :(

**Comment by carmocca** [comment]
Isn't that a feature? PL is a library that fits the criteria of

> What if someone builds a Trainer library with LightningLite:

PL should remove the context manager after this change

**Comment by awaelchli** [comment]
I would have preferred to keep this change of behavior to Lite first. I am worried that there are unintended consequences we haven't thought about.

**Comment by carmocca** [comment]
We will need to face those consequences either way, so I would suggest starting now to uncover issues asap.
